2020-02-Spring/3_trees/recover_bst/java/Solution.java

```diff
@@ -0,0 +1,200 @@
+            inorderTraversal(root.left, nodes);
+            nodes.add(root.val);
+            inorderTraversal(root.right, nodes);
+        }
+    }
+
+    // Helper method that finds the first two values
+    // that are not in ascending order
+    public static int[] findSwappedValues(List<Integer> values) {
+        int length = values.size();
+        int first = -1;
+        int second = -1;
+        for (int i = 0; i < length - 1; i++) {
+            if (values.get(i + 1) < values.get(i)) {
+                if (first == -1) {
+                    first = values.get(i);
+                    second = values.get(i + 1);
+                } else {
+                    second = values.get(i + 1);
+                    break;
+                }
+            }
+        }
+        return new int[]{first, second};
+    }
+    
+    // Helper method that finds the nodes that have the given values
+    // and swaps their values
+    public static void fixSwaps(TreeNode root, int fixes, int first, int second) {
+        if (root != null) {
+            if (root.val == first || root.val == second) {
+                if (root.val == first) {
+                    root.val = second;
+                } else {
+                    root.val = first;
+                }
+                if (--fixes == 0) return;
+            }
+            fixSwaps(root.left, fixes, first, second);
+            fixSwaps(root.right, fixes, first, second);
+        }
+    }
+
+    public static void recoverTree(TreeNode root) {
+        List<Integer> nodes = new ArrayList<>();
+        inorderTraversal(root, nodes);
+        int[] swapped = findSwappedValues(nodes);
+        fixSwaps(root, 2, swapped[0], swapped[1]);
+    }
+
+    /////////////////////////////////////////////////////////
+    ////////////////// OPTIMAL SOLUTION  ////////////////////
+    /////////////////////////////////////////////////////////
+
+    // Helper method for swapping the values of two nodes
+    public static void swap(TreeNode a, TreeNode b) {
+        int temp = a.val;
+        a.val = b.val;
+        b.val = temp;
+    }
+
+    public static void recoverTree2(TreeNode root) {
+        TreeNode first = null;       // First swapped node
+        TreeNode second = null;      // Second swapped node
+        TreeNode predecessor = null; // For Morris traversal
+        TreeNode previous = null;    // For finding the swapped nodes
+        while (root != null) {
+            if (root.left != null) {
+                // Find inorder predecessor of root
+                predecessor = root.left;
+                while (predecessor.right != null && predecessor.right != root) {
+                    predecessor = predecessor.right;
+                }
+                // If there is no link to the parent node
+                // Create one, then visit the left subtree
+                if (predecessor.right == null) {
+                    predecessor.right = root;
+                    root = root.left;
```

Comment:
**Note:** This is a gigantic nitpick that I wouldn't worry about for this particular session, but is something to think about for your own big brain fun times.

It's _probably_ easy to fix, but if you're creating links, you should probably also destroy those links I think. You should be able to do this with something like the following:
```
if not root: return
current = root
if current.left:
    predecessor = current.left
    while predecessor.right:
        predecessor = predecessor.right
    if predecessor.right == current:
        # break the link
        predecessor.right = null
```

This is really rough pseudocode, is certainly incomplete, and is probably incorrect, but should get you in the right direction.